### DISCUSSION TITLE
How to add exclude to an imported matrix

### DISCUSSION THREAD

**Discussion by Cyb3r-Jak3** [community/community#25227] [Actions]
<p>I currently have a workflow from which a matrix in created in a previous step and passed on. I want to add an exclude statement to the matrix, but when I try something like this:</p>
<pre><code class="lang-yaml">    strategy:
      matrix: ${{ fromJson(needs.matrix.outputs.built_matrix) }}
      exclude:
        - bake-target: alpine
          flask: 2.1.0
</code></pre>
<p>I get an error <code>The workflow is not valid. .github/workflows/docker.yml (Line: 40, Col: 7): Unexpected value 'exclude'</code></p>
<p>Is what am I doing possible and if so then how would I accomplish it?</p>

**Comment by ChristopherHX** [comment]
<aside class="quote no-group" data-username="Cyb3r-Jak3" data-post="1" data-topic="241959">
<div class="title">
<div class="quote-controls"></div>
 Cyb3r-Jak3:</div>
<blockquote>
<p>Is what am I doing possible</p>
</blockquote>
</aside>
<p>Yes, if you do it in a differerent way. <strong>This isn’t documented anywhere outside the actions/runner repository.</strong></p>
<pre><code class="lang-yaml">    strategy:
      matrix:
        ${{ include }}: ${{ fromJson(needs.matrix.outputs.built_matrix) }}
        exclude:
          - bake-target: alpine
            flask: 2.1.0
</code></pre>
<p>The mapping key <code>${{ include }}</code> inserts it’s mapping value into the matrix object. However this only works if <code>fromJson(needs.matrix.outputs.built_matrix)</code> doesn’t contain any mapping key named <code>exclude</code>.</p>

**Comment by jsoref** [comment]
<p>Since you’re already using fromJson, you could have a job that constructs the JSON and merges in the excludes and then have the matrix job need the output of the job that defines the JSON.</p>

**Comment by Cyb3r-Jak3** [comment]
<aside class="quote no-group" data-username="ChristopherHX" data-post="2" data-topic="241959">
<div class="title">
<div class="quote-controls"></div>
 ChristopherHX:</div>
<blockquote>
<pre><code class="lang-auto">    strategy:
      matrix:
        ${{ include }}: ${{ fromJson(needs.matrix.outputs.built_matrix) }}
        exclude:
          - bake-target: alpine
            flask: 2.1.0
</code></pre>
</blockquote>
</aside>
<p>When I try that I get <code>Unrecognized named-value: 'include'. Located at position 1 within expression: include</code></p>
<p><a href="https://github.com/Cyb3r-Jak3/pypy-flask/actions/runs/2079922436/workflow" rel="noopener nofollow ugc">Workflow Link</a></p>

**Comment by ChristopherHX** [comment, marked as answer]
<p><a class="mention" href="/cyb3r-jak3">@Cyb3r-Jak3</a> Sorry my fault, it was <code>insert</code>.</p>
<pre><code class="lang-yaml">  Bake:
    needs: matrix
    runs-on: ubuntu-latest
    strategy:
      matrix:
        ${{ insert }}: ${{ fromJson(needs.matrix.outputs.built_matrix) }}
        exclude:
          - bake-target: alpine
            flask: 2.1.0

    steps:
    - run: echo "$MATRIX"
      env:
        MATRIX: ${{tojson(matrix)}}
</code></pre>
<p>Here a syntax checked version, which uses this technique:<br>
<a href="https://github.com/ChristopherHX/newcomposite-sample/actions/runs/2081464240/workflow" rel="noopener nofollow ugc">Workflow Link</a></p>

**Comment by Cyb3r-Jak3** [comment]
<p>Sweet that got it. Thanks!</p>
